Using Elastic's bulk API for upsert operations

sliwinski-milosz opened the pull request:
Finally I prepared tests for BulkBuffer class which uses bulk API for upsert operations.
It has been discussed inside below issue:
https://github.com/mongodb-labs/mongo-connector/issues/446

There is one case which is not handled by this solution:
https://github.com/mongodb-labs/mongo-connector/issues/446#issuecomment-230836737
Does default elastic2_doc_manager handle that case?

I made some small updates to logic in comparison to the version which I published in July - hopefully I haven't messed anything. Tomorrow I will test today's release - in case of issues I will let you know.

Please note that I had to update "test_elastic_commit" unittest as in my opinion it didn't test auto_commit_interval at all. It was passing because after testing auto_commit_interval == None, indexed document even that was removed by _remove function, still was searchable because ES has not been refreshed.

Anyway tests now passes for both: default elastic2_doc_manager and the one with BulkBuffer as well.

I was not able to run tests.test_elastic2.TestElastic because of some problems with starting replica set. I don't think that is related to changes, rather just some my local configuration issues.

ShaneHarvey commented:
Thanks for the pull request! I have one comment before I start reviewing the change. This case https://github.com/mongodb-labs/mongo-connector/issues/446#issuecomment-230836737 is handled by mongo-connector 2.4.1 and elastic2-doc-manager 0.2.0. Perhaps you are testing with outdated versions?

sliwinski-milosz commented:
Yes, you are right. I merged the latest version to my fork.

I am happy to say that I am not able to reproduce the issue which I mentioned.

I noticed that even if refresh is False inside:
`self.elastic.index(index=index, doc_type=doc_type,
                   body=doc, id=doc_id,
                   refresh=(self.auto_commit_interval == 0))`
still document could be visible in Elasticsearch. That is why testing pass for default elastic2_doc_manager version and failed above. I set auto_commit_interval to 0 for testcase which used to fail to immediately commit buffered docs.

The reason why I was not able to run all tests on my local machine was that I tried to run them while mongodb was running - heh - it should be stopped before testing.

BlueRayONE commented:
Any news? Or any possibility, that this PR will be merged and published/released in a near future?
This improvement is happily awaited in my project. Syncing some million documents isn't much fun and takes a lot of time...

ShaneHarvey commented on tests/test_elastic2_doc_manager.py:
```diff
@@ -160,13 +206,18 @@ def test_search(self):
         for autocommit_interval in [None, 0, 1, 2]:
-            docman.auto_commit_interval = autocommit_interval
+            docman = DocManager(elastic_pair,auto_commit_interval=autocommit_interval)
             docman.upsert(docc, *TESTARGS)
```
We definitely need to add more tests for the BulkBuffer besides just a single upsert. At least the cases I mentioned above 
- insert id 1 followed by update id 1
- insert in "test.foo" followed by dropDatabase "test"
- insert in "test.foo" followed by drop "test.foo"

Additionally, can you create a new file `test_bulk_buffer.py` for separate unittests?

sliwinski-milosz replied:
1. Test has been already added
2. Added
3. Added

ShaneHarvey commented on mongo_connector/doc_managers/elastic2_doc_manager.py:
```diff
@@ -342,3 +390,117 @@ def get_last_doc(self):
+        
+        # Dictionary of sources
+        # Format: {"_index": {"_type": {"_id": {"_source": actual_source}}}}
+        self.sources = {}
```
Also, is there any reason you didn't include the `insert_file` indexing in the BulkBuffer?

sliwinski-milosz replied:
I think that `self.commit()` should be called before every command, maybe beside `create` only (updated in code).

For `stop` it should be enough just to change `self.auto_commit_interval = 0` and call `self.commit()`. Whenever something will try to do any update, anyway `self.commit()` will be called.

I didn't include `insert_file` because I don't know much about that operation. It has not been implemented neither by @jmmk who made a first implementation of bulk buffer. There are some more actions done while inserting file (check if elasticsearch threats it as a file, some base64 operations). Can we do updates on inserted file? How can I store that in `self.sources`? If updates can't be done on inserted file, then even that it will be indexed before other operations, it should not make any difference. Correct me if I am wrong.

sliwinski-milosz commented on mongo_connector/doc_managers/elastic2_doc_manager.py:
```diff
@@ -84,14 +86,18 @@ def __init__(self, url, auto_commit_interval=DEFAULT_COMMIT_INTERVAL,
+        
+        self._formatter = DefaultDocumentFormatter()
+        self.BulkBuffer = BulkBuffer(self)
+        self.lock = Lock()
```
But lock is needed outside BulkBuffer as well inside commit method:
```python
    def commit(self):
        """Send bulk requests and clear buffer"""
        with self.lock:
            try:
                action_buffer = self.BulkBuffer.get_buffer()
                if action_buffer:
                    successes, errors = bulk(self.elastic, action_buffer)
            except Exception as e:
                # Handle errors from bulk indexing request
                raise

        retry_until_ok(self.elastic.indices.refresh, index="")
```

We don't want to release lock before bulk to elasticsearch will not be done. It is because we don't want to start another bulk if the first one is still ongoing.

ShaneHarvey replied:
Ah! I see, so the lock is being used to prevent protect access to BulkBuffer and elastic. Can you add a comment to that effect?

sliwinski-milosz commented on mongo_connector/doc_managers/elastic2_doc_manager.py:
```diff
@@ -342,3 +390,117 @@ def get_last_doc(self):
+        # and they are not in local buffer
+        if self.doc_to_get: self.update_sources()
+        
+        ES_buffer = deepcopy(self.action_buffer)
```
Hmm actually all operation which can change `self.action_buffer` are done inside lock. So self.action_buffer could be cleared after bulk to elasticsearch is done.
I just wanted to clear `self.action_buffer` in the same method which returns it for bulk operation.

We can omit that, but we will have to clear the buffer after bulk is done (outside BulkBuffer class). Do you think that it is a better idea?

sliwinski-milosz commented on mongo_connector/doc_managers/elastic2_doc_manager.py:
```diff
@@ -21,10 +21,12 @@
 import warnings
 
-from threading import Timer
+from threading import Timer,Lock
```
Done

sliwinski-milosz commented on mongo_connector/doc_managers/elastic2_doc_manager.py:
```diff
@@ -310,13 +341,30 @@ def search(self, start_ts, end_ts):
+            try:
+                action_buffer = self.BulkBuffer.get_buffer()
+                if action_buffer:
+                    successes, errors = bulk(self.elastic, action_buffer)
```
Done.
I think that try/expect here is not needed actually.

sliwinski-milosz commented on mongo_connector/doc_managers/elastic2_doc_manager.py:
```diff
@@ -342,3 +390,117 @@ def get_last_doc(self):
+        if current_type:
+            self.sources[index][doc_type][document_id] = new_source
+        else:
+            self.sources.setdefault(index,{})[doc_type] = {document_id : new_source}
```
Yes, you are right. Done.

sliwinski-milosz commented:
@ShaneHarvey thank you very much for deep review. I made fixes for requested changes.

For some reason testing failed in travis. Please note that all tests pass on my local machine.
Seems like there is some problem with travis configuration, am I right?
`error: Setup script exited with error in urllib3 setup command: Invalid environment marker: python_version <= "2.7"`

ShaneHarvey commented on mongo_connector/doc_managers/elastic2_doc_manager.py:
```diff
@@ -342,3 +390,117 @@ def get_last_doc(self):
+                self.doc_to_get = []
+                raise errors.OperationFailed(
+                    "mGET: Document id: {} has not been found".format(each_doc['_id']))
+        self.doc_to_get = []
```
I was confused about when `add_to_sources` was called. Let me go through this another time to see if we can make the code clearer.

sliwinski-milosz commented:
I have an idea how to handle `update_sources` in the little bit clearer way, and how to get source from elasticsearch only for documents which we really need. I will implement that on Saturday.

@ShaneHarvey could you please elaborate little bit more about unnecessary `deepcopy` of `action_buffer`? :) I don't get your point. When should we return it and when should we clear it?

I will also add `insert_file` to the buffer as well.

ShaneHarvey commented:
That sounds great. What I meant is that `get_buffer` can return and clear `self.action_buffer` without using deepcopy:
```python
    def get_buffer(self):
        """Get buffer which needs to be bulked to elasticsearch"""
        
        # Get sources for documents which are in elasticsearch
        # and they are not in local buffer
        if self.doc_to_get:
            self.update_sources()
        
        action_buffer = self.action_buffer
        self.action_buffer = []
        self.sources = {}
        return action_buffer
```

sliwinski-milosz commented:
ou actually you are right as `self.action_buffer = []` dereference the buffer list from `self.action_buffer` but reference is being held by new variable.
I will fix that as well. Thanks!

sliwinski-milosz commented:
Hi Guys,
again some updates from me:

1. Unnecessary `deepcopy` call removed.
2. `insert_file` moved into buffer.
3. I removed refreshing Elasticsearch just before calling mGet API as I found that Get API by default works in realtime ([Get API documentation](https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-get.html#realtime)). My assumption is that mGet API works the same way. To be sure I even added one more unittest to test that -> and it works as expected.
4. And the biggest change. I added some improvements for BulkBuffer. With them, it will get only once the source from Elasticsearch even if there will be more updates to the same document.
I used similar way as for storing sources locally -> which means that I created new dictionary, in which I mark if document is already marked for retrieving from Elasticsearch. You may ask why I store that information twice. It is because of performance. Let me explain:
First place where I store that information is `self.doc_to_update` list. And as it is a list, then it is hard to get that information whenever I will want to check if some specific doc_id is there. To get that I would have to loop through whole list. That is why I store information in dictionary as well.

llvtt commented on mongo_connector/doc_managers/elastic2_doc_manager.py:
```diff
@@ -333,13 +391,30 @@ def search(self, start_ts, end_ts):
     def run_auto_commit(self):
         """Periodically commit to the Elastic server."""
-        self.elastic.indices.refresh()
+        self.commit()
```
I agree. OplogThread will continue even if an uncaught Exception is raised here, but the automatic commit functionality will have stopped. `commit()` should probably be inside a try/except.

sliwinski-milosz replied:
I put `commit` into try/except and I log exceptions.

ShaneHarvey replied:
Great!

llvtt commented on mongo_connector/doc_managers/elastic2_doc_manager.py:
```diff
@@ -365,3 +440,163 @@ def get_last_doc(self):
+            documents = self.docman.elastic.mget(body={'docs': docs})
+            return iter(documents['docs'])
+        else:
+            return []
```
I haven't read this entire change set, but we're returning a `list` in one case and an iterator in another. They don't have the same interface, so this could break things that depend on having an iterator:

```python
>>> next([])
TypeError: list object is not an iterator
```

sliwinski-milosz replied:
There is no such case that it will call next on that if it is empty. But anyway I changed that to `iter([])` for consistency :) Thanks.

llvtt commented on mongo_connector/doc_managers/elastic2_doc_manager.py:
```diff
@@ -439,75 +472,102 @@ def __init__(self, docman):
+                else:
+                    # Document not found in elasticsearch,
+                    # Seems like something went wrong during replication
+                    self.doc_to_get = []
```
If the particular failure here doesn't imply that we can't continue applying the other operations in `doc_to_update`, then probably continuing is best. This is the same approach that other doc managers take when applying updates: if the update cannot be applied, then raise an Exception to let the OplogThread know. OTOH, if the further operations in `doc_to_update` don't make any sense if a previous one failed, then you may as well just raise an Exception here as per now.

sliwinski-milosz replied:
After changes I log the errors instead of raising Exception.

sliwinski-milosz commented:
Thank you for another review. I will fix all mentioned things on Saturday.

sliwinski-milosz commented on mongo_connector/doc_managers/elastic2_doc_manager.py:
```diff
@@ -439,75 +472,102 @@ def __init__(self, docman):
+            return False
+        else:
+            self.set_doc_to_get_id(action)
+            return True
```
Yes, that is a good idea. Done.

sliwinski-milosz commented on tests/test_elastic2_doc_manager.py:
```diff
@@ -251,6 +357,87 @@ def test_commands(self):
+        self.elastic_doc.upsert(doc, *cmd_args)
+
+        self.elastic_doc.handle_command({'drop': doc_type}, *cmd_args)
+        time.sleep(3)
```
I copied `sleep` from other tests :).
Today replaced `sleep` in all of them where it has been used for the same purpose.

benan789 commented:
Do we have an ETA on this?

ShaneHarvey commented:
Merged in https://github.com/mongodb-labs/elastic2-doc-manager/commit/5ed370c61ef13ad0b8b27ffb3b108736eb875f39.

BlueRayONE commented:
Awesome! Big thanks to @sliwinski-milosz for this great work, I waited a long time for this feature.

I directly tested it and deployed the new dev-version on my project (nearly 80 million documents).
Firstly, I can confirm a huge performance-boost! 

But I also noticed, that not everything is synced properly. On every insert I have a "small" number of missing documents between mongoDB and ES, which is not synced. I observed differences between 70 and 900 documents. 
I can see which documents are missing. When inserting new ones, the old ones - which where missing before - are synced then. That brings me to the idea, that maybe the **last** collection of new documents, which is collected for the next bulk-insert, is perhaps not transferred/committed properly to ES?

Otherwise, I have no idea for a solution. My logfiles are clean. (Ok, my mongoDB-log is not informative due to [this circumstance](https://jira.mongodb.org/browse/SERVER-1220)).

Additional information: I'm using mongoDB 3.2.10 and ES 2.4.1, both single nodes.

By the way, I don't know if it would be better to open a new issue for this? Or handle this here because it's directly related to this PR?

sliwinski-milosz commented:
@BlueRayONE 
For me it looks like it is not an issue with BulkBuffer - it sounds like it is the problem with yours configuration.

Do you have configured [autoCommitInterval](https://github.com/mongodb-labs/mongo-connector/wiki/Configuration-Options#docmanagersindexautocommitinterval)? If not, it just means that elastic2_doc_manager commits changes only after bulk is filled (by default 1000 docs). And documents, which are missing are the ones which are stored in bulk and waiting till it will be filled.

Try to set yours autoCommitInterval to 2 -> thanks to that it will commit docs from bulk and clean it every 2 secs.

BlueRayONE commented:
Oh, of course. Makes totally sense.
I changed it (connector restart included), but the difference is still there. A resync is now neccessary, right? The already partially filled bulk is lost due to the restart, but the oplog-timestamp is already set to the newest entry. So the difference would never be synced without resync.
I'll come back when resync is done (takes some time with my 80 million documents...).

Thanks.

sliwinski-milosz commented:
AFAIK resync shouldn't be necessary but only in case that you gracefully stopped the mongo-connector. If you killed it, then of course you have to resync.

There is a logic in code that will commit docs from bulk which stopping mongo-connector.

ShaneHarvey commented:
@BlueRayONE Can you create a new issue for this? Please include the information above and we'll take it from there.

BlueRayONE commented:
Again a small feedback for @sliwinski-milosz: Everything is running stable until now, having thousands of new documents everyday. Speed is also fast as hell everytime ;)
